drivers/vhost/dsp-sof.c

```diff
@@ -0,0 +1,968 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2017-2019 Intel Corporation. All rights reserved.
+ *
+ * Author:	Libin Yang <libin.yang@intel.com>
+ *		Luo Xionghu <xionghu.luo@intel.com>
+ *		Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *		Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+#include <asm/unaligned.h>
+
+#include <linux/device.h>
+#include <linux/file.h>
+#include <linux/firmware.h>
+#include <linux/fs.h>
+#include <linux/hw_random.h>
+#include <linux/miscdevice.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/uio.h>
+#include <linux/vmalloc.h>
+#include <linux/workqueue.h>
+
+#include <sound/pcm_params.h>
+#include <sound/sof.h>
+
+#include <sound/sof/virtio.h>
+
+#include "dsp.h"
+
+#include "../sound/soc/sof/sof-priv.h"
```

Comment:
This file is mostly audio code so it should reside in sound/soc/sof, we can export functions for vhost/dsp.c to use though.

Reply:
@lgirdwood but this code is only used for vhost, so, no need to link it into snd-sof, right? At the moment I build a vhost-dsp module under drivers/vhost, upon which snd-sof depends, because it calls some functions from it. So, we cannot link this file with snd-sof, that would create a cross-dependency. Do you want that as a separate module? In general I think drivers/vhost/ does contain code, specific to respective subsystems like scsi etc. Maybe *in principle* it's also possible to build vhost drivers outside of drivers/vhost, but presumably that would be against the convention.

Reply:
good luck upstreaming between 2 subsystems.

Reply:
I think it's possible to split the patch into two for SOF and for vhost. The sound part will prepare all the interfaces, but will work even without the vhost part, because without the CONFIG_VHOST_SOF defined all vhost functions will be replaced by dummies. After that part is in the mainline, we can push vhost too.